File: livingrimoire start here/LivinGrimoire java/LivinGrimoirePacket/RailPunk/EventChatV2.java
package LivinGrimoirePacket.RailPunk;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

class EventChatV2 {
    private final int lim;
    private final Map<String, WeightedResponder> dic = new HashMap<>();
    private final Set<String> modifiedKeys = new HashSet<>();

    EventChatV2(int lim) {
        this.lim = lim;
    }

    Set<String> getModifiedKeys() {
        Set<String> replica = new HashSet<>(modifiedKeys);
        modifiedKeys.clear();
        return replica;
    }

    boolean keyExists(String key) {
        return modifiedKeys.contains(key);
    }

    void addFromDb(String key, String value) {
        if (value == null || value.isEmpty() || value.equals("null")) {
            return;
        }
        String[] values = value.split("_");
        dic.computeIfAbsent(key, k -> new WeightedResponder(lim));
        for (String item : values) {
            dic.get(key).addResponse(item);
        }
    }

    void addKeyValue(String key, String value) {
        modifiedKeys.add(key);
        dic.computeIfAbsent(key, k -> new WeightedResponder(lim)).addResponse(value);
    }

    void addKeyValues(List<AXKeyValuePair> pairs) {
        for (AXKeyValuePair pair : pairs) {
            addKeyValue(pair.getKey(), pair.getValue());
        }
    }

    String response(String in1) {
        WeightedResponder wr = dic.get(in1);
        return wr != null ? wr.getAResponse() : "";
    }

    String responseLatest(String in1) {
        WeightedResponder wr = dic.get(in1);
        return wr != null ? wr.getLastItem() : "";
    }

    String getSaveStr(String key) {
        WeightedResponder wr = dic.get(key);
        return wr != null ? wr.getSavableStr() : "";
    }
}

File: livingrimoire start here/LivinGrimoire Kotlin/auxiliary_modules/EventChatV2.kt
package auxiliary_modules

import java.util.HashMap
import java.util.HashSet

class EventChatV2    // Constructor
    (private val lim: Int) {
    //    private final Map<String, LimUniqueResponder> dic = new HashMap<>(); // old version
    private val dic: MutableMap<String, WeightedResponder> = HashMap()
    val modifiedKeys = HashSet<String>()

    fun keyExists(key: String): Boolean {
        // if the key was active true is returned
        return modifiedKeys.contains(key)
    }

    // Add items
    fun addItems(ur: WeightedResponder, vararg args: String) {
        for (arg in args) {
            dic[arg] = ur.cloneObj()
        }
    }

    fun addFromDB(key: String, value: String) {
        if (value.isEmpty() || value == "null") {
            return
        }
        val tool1 = AXStringSplit()
        val values = tool1.split(value)
        if (!dic.containsKey(key)) {
            dic[key] = WeightedResponder(lim)
        }
        for (item in values) {
            dic[key]!!.addResponse(item)
        }
    }

    // Add key-value pair
    fun addKeyValue(key: String, value: String?) {
        modifiedKeys.add(key)
        if (dic.containsKey(key)) {
            dic[key]!!.addResponse(value!!)
        } else {
            dic[key] = WeightedResponder(lim)
            dic[key]!!.addResponse(value!!)
        }
    }

    fun addKeyValues(elizaResults: ArrayList<AXKeyValuePair>) {
        for (pair in elizaResults) {
            // Access the key and value of each AXKeyValuePair object
            addKeyValue(pair.key, pair.value)
        }
    }

    // Get response
    fun response(in1: String): String {
        return if (dic.containsKey(in1)) dic[in1]!!.aResponse else ""
    }

    fun responseLatest(in1: String): String {
        return if (dic.containsKey(in1)) dic[in1]!!.lastItem else ""
    }

    fun getSaveStr(key: String): String {
        return dic[key]!!.savableStr
    }
}